### PR TITLE
[HOPSWORKS-1459] provenance support - file and app

### PR DESCRIPTION
**Pull request by o-alex**
jira - https://logicalclocks.atlassian.net/browse/HOPSWORKS-1459

add companion table handling
index per project
prov xattrs
xattr delete op
deal with composite key - attr_prov.key
bulk can include index and type in the body of the request
hive/featurestore

### CODE REVIEW

File: include/ElasticSearchBase.h [hunk 1]
```diff
@@ -34,12 +34,13 @@ class ElasticSearchBase : public TimedRestBatcher, public
 
 protected:
   std::string getElasticSearchUrlonIndex(std::string index);
-  std::string getElasticSearchUrlOnDoc(std::string index, Int64 doc);
-  std::string getElasticSearchUpdateDocUrl(std::string index, Int64 doc);
-  std::string getElasticSearchBulkUrl(std::string index);
+  std::string getElasticSearchUrlOnDoc(std::string index, std::string docType, Int64 doc);
+  std::string getElasticSearchUpdateDocUrl(std::string index, std::string docType, Int64 doc);
+  std::string getElasticSearchBulkUrl(std::string index, std::string docType);
```
Comment (maismail) [inline include/ElasticSearchBase.h]: Why did you add a docType as parameter? in Elasticsearch they removed the document types and right now they keep only _doc for compatibility not used as type but just part of the request url and in newer versions they are going to remove it.

File: include/FileProvenanceElastic.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  FileProvenanceElastic(std::string elastic_addr, std::string index,
+          int time_to_wait_before_inserting, int bulk_size,
+          const bool stats, SConn conn);
+
+  virtual ~FileProvenanceElastic();
+private:
+  const std::string mIndex;
+  SConn mConn;
+
+  const FileProvenanceLogTable::FileProvLogHandler* getLogHandler(const LogHandler* log);
```
Comment (maismail) [inline include/FileProvenanceElastic.h]: why do you need this method?

File: include/tables/XAttrTable.h [hunk 3]
```diff
@@ -124,6 +124,8 @@ struct XAttrRow {
     opWriter.EndObject();
 
     opWriter.EndObject();
+
+    LOG_INFO("script:" << sbOp.GetString());
```
Comment (maismail) [inline include/tables/XAttrTable.h]: remove the log or if you need to keep it just write a more meaningful one

File: src/AppProvenanceElastic.cpp [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+      if (mStats) {
+        mCounters.bulkProcessed(start_time, bulk);
+      }
+    }
+  }
+}
+
+bool AppProvenanceElastic::bulkRequest(eEvent& event) {
+  if (httpPostRequest(mElasticBulkAddr, event.getJSON())){
+    if(event.getLogHandler()->getType() == LogType::PROVAPPLOG){
```
Comment (maismail) [inline src/AppProvenanceElastic.cpp]: no need for the check since you are only handling one log table in this class

File: src/ElasticSearchBase.cpp [hunk 5]
```diff
@@ -28,19 +28,24 @@ std::string ElasticSearchBase::getElasticSearchUrlonIndex(std::string index) {
   return str;
 }
 
-std::string ElasticSearchBase::getElasticSearchUrlOnDoc(std::string index, Int64 doc) {
+std::string ElasticSearchBase::getElasticSearchUrlOnDoc(std::string index, std::string docType, Int64 doc) {
   std::stringstream out;
-  out << getElasticSearchUrlonIndex(index) << "/" << DEFAULT_TYPE << "/" << doc;
+  out << getElasticSearchUrlonIndex(index) << "/" << docType << "/" << doc;
```
Comment (maismail) [inline src/ElasticSearchBase.cpp]: same as before we should keep the DEFAULT_TYPE and don't add it as a parameter

File: src/FileProvenanceElastic.cpp [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+#include <FileProvenanceConstants.h>
+#include "FileProvenanceElastic.h"
+
+FileProvenanceElastic::FileProvenanceElastic(std::string elastic_addr, std::string index,
+        int time_to_wait_before_inserting, int bulk_size, const bool stats, SConn conn) :
+ElasticSearchWithMetrics(elastic_addr, time_to_wait_before_inserting, bulk_size, stats),
+mIndex(index), mConn(conn) {}
+
+const FileProvenanceLogTable::FileProvLogHandler* FileProvenanceElastic::getLogHandler(const LogHandler* log) {
```
Comment (maismail) [inline src/FileProvenanceElastic.cpp]: same as before, do you use this method, you shouldn't have it here

File: src/main.cpp [hunk 7]
```diff
@@ -45,7 +45,8 @@ int main(int argc, char** argv) {
     int elastic_batch_size = 5000;
     int elastic_issue_time = 5000;
 
-    std::string elastic_provenance_index = "provenance";
+    std::string elastic_file_provenance_index = "";
```
Comment (maismail) [inline src/main.cpp]: what will be the file provenance index, was it per project? is this parameter even needed?

File: src/ProjectsElasticSearch.cpp [hunk 8]
```diff
@@ -89,22 +89,22 @@ bool ProjectsElasticSearch::bulkRequest(eEvent& event) {
 }
 
 void ProjectsElasticSearch::addDoc(Int64 inodeId, std::string json) {
-  std::string url = getElasticSearchUpdateDocUrl(mIndex, inodeId);
+  std::string url = getElasticSearchUpdateDocUrl(mIndex, DEFAULT_TYPE, inodeId);
```
Comment (maismail) [inline src/ProjectsElasticSearch.cpp]: same as before the type issue, it is not limited to ProjectsElasticSearch

File: include/ElasticSearchBase.h (hunk 1 above)
Comment (o-alex) [reply include/ElasticSearchBase.h]: yeah i changed one of my doc types...
fixed it now... removed

File: include/FileProvenanceElastic.h (hunk 2 above)
Comment (o-alex) [reply include/FileProvenanceElastic.h]: removed ... old code

File: include/tables/XAttrTable.h (hunk 3 above)
Comment (o-alex) [reply include/tables/XAttrTable.h]: fixed - it was leftover debug code

File: src/AppProvenanceElastic.cpp (hunk 4 above)
Comment (o-alex) [reply src/AppProvenanceElastic.cpp]: fixed

File: src/ElasticSearchBase.cpp (hunk 5 above)
Comment (o-alex) [reply src/ElasticSearchBase.cpp]: fixed

File: src/FileProvenanceElastic.cpp (hunk 6 above)
Comment (o-alex) [reply src/FileProvenanceElastic.cpp]: done

File: src/main.cpp (hunk 7 above)
Comment (o-alex) [reply src/main.cpp]: fixed - removed the index

File: src/ProjectsElasticSearch.cpp (hunk 8 above)
Comment (o-alex) [reply src/ProjectsElasticSearch.cpp]: fixed
